### PR TITLE
Finalizing changelog for v4.3rc1

### PR DESCRIPTION
**Pull request by eteq**
This PR is an attempt to make up the correct workflow to sync up the towncrier changes for 4.3.x with main (see #11847 as well).

I actually did the towncrier command in 4.3.x as part fo the release process (8ea8876d375a7105308e695e881330487924a154) but then cherry-picked the SHA to make this PR.  It seems to have gone smoothly - should I document this as the recommended workflow, @saimn @pllim @bsipocz @astrofrog?

My one question though, is: should we merge this *now* with the RC, or should we wait until the final 4.3 release happens? I'm not really clear on what we want to use as the workflow for anything that needs to be included between the RC and the real release... We can't use this *exact* same process because it would then have two separate sections, one for 4.3rc1 and another for 4.3.  Does towncrier have a way to clean that up?  Or should I just do a "manual" commit at 4.3 release that by-hand merges any post-RC changelog entries and updates the date/version name?

### CODE REVIEW

**Comment by eteq** [comment]
Wow the labeling bot went a bit nuts with this one.  Not inaccurate, but kinda amusing!

**Comment by pllim** [comment]
I have to recuse myself from this one. I still don't grok towncrier. @saimn is the best person to ask.

File: CHANGES.rst
Comment (bsipocz) [inline CHANGES.rst:107]: reformat, so no lines go over the 80 chars, some of the ones above may also be longer than 80.

TODO: we should do a pre-commit that cuts the towncrier snippets into multi lines, as people will be always run over otherwise

File: CHANGES.rst
Comment (bsipocz) [inline CHANGES.rst:35]: Some of these changelog entries are unusually long, maybe it's a side effect of towncrier? Overall I would reemphasize that these should be much briefer, keep the details for the what's new.

File: CHANGES.rst
Comment (bsipocz) [inline CHANGES.rst:1]: fix this

File: CHANGES.rst
Comment (bsipocz) [inline CHANGES.rst:136]: Either brief this, or cut up to multiple entries.

File: CHANGES.rst
Comment (bsipocz) [inline CHANGES.rst:228]: break it into two entries rather than use "Also" in the same entry

File: CHANGES.rst
Comment (bsipocz) [inline CHANGES.rst:299]: Used to have two empty lines between the 3 main sections

File: CHANGES.rst
Comment (bsipocz) [inline CHANGES.rst:497]: Other changes and additions should not have module specific entries, this section is really just for the extra stuff

**Review by bsipocz — CHANGES_REQUESTED** [review]
The RC certainly should have a changelog shipped with it. But! I would keep the previous approach of updating the heading from 4.3.rc1 to 4.3 for the release, so for the actual release there won't be an RC entry.
(maybe that means that we should not delete the snippets? I don't know, my take is that once the changelog is rendered they are not relevant anymore and should be deleted).

I left a couple of comments, it seems that we have some minor formatting issues that maybe a pre-commit could resolve easily for contributors, but otherwise we just need to keep doing manual cleanups.
The most important that the 4th section, "Other changes" is not right, it should not have any module level content.

Also, I noticed that entries got much much longer with narrative words that brings no meaning, I much more like the terser approach we had before, keep the fluffy descriptive sentences into the what's new and into the description of the PR for those who care enough to look under the hood (after all we list the PR number in the changelog for a reason).

**Comment by saimn** [comment]
Since we don't keep a separate changelog for rc releases in the final release, I guess what I would do is:
- for rc releases, generate the changelog but don't delete the fragments.
- when doing the final release, delete the rc section and generate agin the changelog, delete the fragments, and backport this.
But there may be other ways to do it.

**Comment by saimn** [comment]
About fixing typos, yeah this is a bit annoying, not sure which is the best way to proceed. I don't understand your solution (prior tagging RC etc.) ?
For 4.3.rc1, Erik rendered the changelog on the release branch and this PR is the "forward port", if I understood correctly.

**Comment by bsipocz** [comment]
> For 4.3.rc1, Erik rendered the changelog on the release branch and this PR is the "forward port", if I understood correctly.

Then it should not be milestoned to 4.3. Also, there are a ton of issues with it, e.g it's not exactly the same version that went in the RC as a few fixes have been done there, and we suggest a few fixes here, and we likely will end up with divergent versions on the two branches. 
So, if it's a draft that's open to review then wait for that review, but if it goes to production then merge away to avoid diverging versions.

**Comment by eteq** [comment]
> But maybe a solution where the changelog is getting done and fixed up on the release branch, including deleting the fragments should be done prior tagging RC, but the forward porting of it would only happen at the actual release, so the main branch still has the fragments up until the official release is out?

Yeah, I mostly like this. But to limit the number of back-and-forth PRs, I suggest a PR like this one as the place where review can be done between when the RC is out and the final release (that's was sort of my "second" option in the OP). 

So to be concrete, this is what I'm thinking should be the process:
1. As part of the *RC* release process, the releaser runs the towncrier script, and then does whatever manual cleanup is needed at the time, with the heading `4.3rc1 (date)`
2. As part of the RC process, they create a PR like this one back into `main`. But *unlike* this one, it should be labeled `4.3 (unreleased)`.  Note that this is a bit awkward because it requires copying stuff between v4.3.x and a PR branch, but I don't really see any way to avoid that and still have a useful reviewable PR.
3. Between the RC and the release (or next RC), the community can review the PR and correct anything that needs correcting as @bsipocz did here.
4. If there's multiple RCs with new fragments needed, the releaser repeats 1-3 (but they should *update* the PR with new changelog entries, not make a new one in step 2)
5. Once the release is ready, the releaser runs the script once more if there are any fragments and pushes that up to the PR
6. Thew releaser updates the date, merges the PR, and as the last commit before release, they cherry-pick the merged version back into the release branch to synchronize everything.

So as not to delay 4.3 further I'm going to follow this process for *this* PR (i.e., accept @bsipocz's comments - thanks for the review! - and then backport), with a manual fixup step to work around the fact that I didn't follow 1-3 exactly (fortunately nothing backported after the rc seems to have a changelog entry).  But if we want to discuss further, we can modify it in the future! We'll need to write down the answer in the docs a la #11847, regardless, though.

File: CHANGES.rst
Comment (eteq) [reply CHANGES.rst:497]: Yeah, that's weird.  I don't really understand how town crier decided that, but I will just manually fix it for now.

File: CHANGES.rst
Comment (eteq) [reply CHANGES.rst:299]: Thanks, will fix

Comment (eteq) [reply CHANGES.rst:299]: Actually, I see where to do this in the template, so I'll just do that at the same time.

File: CHANGES.rst
Comment (saimn) [reply CHANGES.rst:497]: towncrier doesn't decide ;), it's up to people to put the fragment in the correct location or to a reviewer/bot to check that :grimacing: 

So fragments for "Other changes and additions" should go in `docs/changes/`, and if they are put in a sub-directory they will be rendered in a sub-category.

```
❯ ls -1 docs/changes/*/*other*
docs/changes/coordinates/11069.other.rst
docs/changes/coordinates/11073.other.rst
docs/changes/coordinates/11090.other.rst
docs/changes/coordinates/11449.other.rst
docs/changes/coordinates/11485.other.rst
docs/changes/coordinates/11601.other.rst
docs/changes/coordinates/11753.other.rst
docs/changes/cosmology/11530.other.rst
docs/changes/io.ascii/11919.other.rst
docs/changes/nddata/11107.other.rst
docs/changes/nddata/11553.other.rst
docs/changes/stats/11219.other.rst
docs/changes/stats/11356.other.rst
docs/changes/utils/11490.other.rst
```

Comment (saimn) [reply CHANGES.rst:497]: It seems `other` was chosen a lot for changes that looked neither like a bugfix nor like a feature, but not sure where those changes would have belong before ?

Comment (bsipocz) [reply CHANGES.rst:497]: there is no such thing as not a bugfix nor a feature. Then those are refactoring and thus internal and has no place in the changelog 🤷 I suspect this really is just a side effect of not reviewing those entries/name of files such as it was the practice before (e.g. I don't think mainainers would have let an entry go through in "Other" category before, and also people didn't think it's a category as it wasn't prepopulated with module names). So, we just need to clean up now at release.

Also, @eteq, keep in mind that with towncrier there are no examples provided for contributors for how a changelog entry looks like, so there is a much large spread of quantity and quality, thus more manual cleanup is needed at release time. Which I think is OK, we minimized on contributor convenience and PR conflicts not release easiness. So, the release manager certainly need to step in a clean up much more than before (note: there was quite a bit of cleanup to he changelog before, too.)

File: CHANGES.rst
Comment (eteq) [reply CHANGES.rst:35]: Yeah, a good point. I'm guessing it's the psychology that you don't see a bunch of other bullets that are short and to the point?  So yeah, emphasizing briefer is a good idea.  Maybe we should update the PR guidelines for this or something?

File: CHANGES.rst
Comment (eteq) [reply CHANGES.rst:228]: After thinking about how to do that I realized the "also" is really an implementation detail anyway.

File: CHANGES.rst
Comment (eteq) [reply CHANGES.rst]: (I fixed these cases by hand for now but I agree it's a great idea to automate this, so not resolving this one)

File: CHANGES.rst
Comment (eteq) [reply CHANGES.rst:522]: Hmm, this is a bigger topic, methinks - should probably discuss it with the wider infrastructure group. 

But I think it is *not* desirable to increase the release manager workload in this context. After all, this is already a bottleneck and we did all the work with the backport bot to simplify the release process by having the author fix this sort of thing. So it sounds like we're trading the work of backporting for the work of changelog cleanup. But I think we can ameliorate this by pointing contributors to some examples.  I mean, we could even have a note in the docs or the PR template or somewhere else that says "look in the changelog in these sections to see how this would render"

That said, I just went through all of the "others", and they all looked like they belonged as *features* because they were "refactor to speed up".  The feature there is really the "speed up", but it's still useful to have in the changelog.  Or put another way, those are things we've had in the changelog in the past.  But they were always "new features".

Comment (eteq) [reply CHANGES.rst:522]: (for 4.3 I'm just going ahead and manually doing it as per the above, but I think we want to make that simpler in the long run since it's potentially quite error prone.)

Comment (eteq) [reply CHANGES.rst:522]: we may want to trace a few of those "other" cases and find out how those contributors first found there way there, since that will let us see where we could have intervened to put them on the happier path

**Comment by eteq** [comment]
Haha, the github action is unhappy that I'm editing the changelog even though I marked it `no-chagelog-entry-needed` :zany_face: .  I think these are a special case that can be ignored though.

**Comment by eteq** [comment]
@bsipocz, can you have a quick look at the updates and approve if this looks good to you now?

**Comment by pllim** [comment]
Ops... `Changelog exists when "no-changelog-entry-needed" label is set`

File: CHANGES.rst
Comment (bsipocz) [reply CHANGES.rst:522]: the tradeoff is days worth of work in terms of backport, and nitpicking on wordsmithing for the changelog. I don't think the latter should be a bottleneck, but of course would be nice if maintainers were picky at the time of PR opening to catch issues. But I don't think it's a have a cake and eat is situation, without rendering and seeing (not just listing in the dev docs) of good, compact examples people will just write long fluffy sentences that are not following those guidelines. 

Again, with towncrier we traded the conflicting PRs due to the changelog, the task to figure out the right version number section etc for someone to have a final look and cleanup at the end. It's a good tradeoff, could be better if maintainers are more nitpicky, but not a catastrophic situation to be in either.

**Comment by bsipocz** [comment]
> Haha, the github action is unhappy that I'm editing the changelog even though I marked it no-chagelog-entry-needed 🤪 . I think these are a special case that can be ignored though.

we used to have a label to ignore those PRs. At least this was worked around before

File: CHANGES.rst
Comment (bsipocz) [inline CHANGES.rst:110]:
```suggestion
- The ``convolve_models_fft`` function implements model convolution so that one
 insures that the convolution remains consistent across multiple different
 inputs. [#11456]
```

File: CHANGES.rst
Comment (bsipocz) [inline CHANGES.rst:58]:
```suggestion
- Added support for reading and writing ASCII tables in QDP (Quick and Dandy
 Plotter) format. [#11256]
```

File: CHANGES.rst
Comment (bsipocz) [inline CHANGES.rst:386]:
```suggestion
```

**Comment by pllim** [comment]
> we used to have a label to ignore those PRs

We did with `astropy-bot`. It wasn't in the towncrier version, but then again, that version isn't working, so I am controling the Action version of it... 😉 I added the logic upstream. It passed now.

File: CHANGES.rst
Comment (bsipocz) [inline CHANGES.rst:536]: drop the numpydoc link, it would and will eventually break and doesn't bring much info into the changelog (which traditionally thus doesn't have links...)

File: CHANGES.rst
Comment (bsipocz) [inline CHANGES.rst:530]: Officially? Nitpick, but I like to keep a standard text for any version support drop, much easier to find and search.

Comment (bsipocz) [reply CHANGES.rst:530]: (and this is not consistent so far, but there are two main formats, we should pick one and stick to it:
`Versions of xyz  <a.b are no longer supported.` or `Minimum version of required xyz is now a.b.`

**Review by bsipocz — APPROVED** [review]
more nitpicking of long lines (there maybe a few more I haven't notice, and a few inconsistencies, but overall this is now OK).

File: CHANGES.rst
Comment (saimn) [reply CHANGES.rst:522]: Agreed, the drawbacks we see here are a bit annoying but that remains a minor annoyance compared to the benefits. Now I guess the question is how to improve this for future releases, it would be great if we could fix the bot, also maybe we can add some checks with `pre-commit`.

**Comment by eteq** [comment]
had to rebase to integrate one more late-breaking backport needed to fix the CI in 4.3.x (#11921)

File: CHANGES.rst
Comment (bsipocz) [inline CHANGES.rst:9612]: Why are these here?

Comment (eteq) [reply CHANGES.rst:9612]: I'm not sure but I bet it was due to rebase getting very confused?  Checking now, thanks for catching that.

File: CHANGES.rst
Comment (eteq) [reply CHANGES.rst:528]: Right, to be clear I was not arguing against towncrier, just that increasing the workload on the release managers was one of the things we were trying to not do.  But I agree there's a have cake and eat it too solution, just have to decide which!

Comment (saimn) [reply CHANGES.rst:528]: Yep. Just to mention it here,  #11945 adds a `pre-commit` that would check those towncrier filenames and avoid this issue in the future.

File: CHANGES.rst
Comment (eteq) [reply CHANGES.rst:9612]: No, never mind, this was a my-bad in e92f9cf7c92db8e0d18b5404cc387e458daedf3b due to getting lost in the changelog.  Will fix in a commit momentarily

Comment (eteq) [reply CHANGES.rst]: OK should be fixed now.  Thanks for catching that @bsipocz !

File: CHANGES.rst
Comment (pllim) [inline CHANGES.rst:1]: Do you have to insert a real date here now, or is that for later?

Comment (pllim) [reply CHANGES.rst:1]: Aside from this question, https://astropy--11856.org.readthedocs.build/en/11856/changelog.html#unreleased seems to render fine, though I did not actually check to see if every entry is where they are supposed to be; I hope one of those release scripts have already taken care of that.

Comment (bsipocz) [reply CHANGES.rst:1]: Also, could you squash all commits together at this point? It's starting to get very incremental.

Comment (eteq) [reply CHANGES.rst:1]: I couldn't find a way to flatten it down to one that didn't end up removing some context for interpreting this PR thread, but I got it down to 5 at least, removing all the repetitve ones.

Comment (eteq) [reply CHANGES.rst:1]: Oh, and re: the "real date", that goes *after* the release, since you never know when there will be a delay that pushes the date forward...

**Review by pllim — APPROVED** [review]
LGTM. Thanks!

p.s. One thing I don't like about towncrier stuff is that we still have to make sure the rendered change log is correct at release time. In the before times, more work is needed at the PR stage but by the time release is happening, you don't have to worry about this part. 🤷

**Pull request merged**
